features/332-sso/api.md

```diff
@@ -0,0 +1,311 @@
+
+> In addition to email, should we consider phone number?
+
+# Removal of v1 API
+
+The following v1 API is removed.
+
+```typescript
+function getOAuthProviderProfiles(): Promise<{[providerID: string]: any}>;
+```
+
+The existing implementation simply return all snapshots of user info of all providers of the logged in user. The snapshot of user info most likely will be stale. So the function unlikely return what the developer expects. If we want to make the function always return the fresh user info, we need to implement refresh token of OAuth 2.0. However, to implement refresh token, we have to drop support for `withAccessToken`, not to mention that some providers like Facebook does not support refresh token.
+
+# List of Supported OAuth 2.0 Providers
+
+- Google
+- Facebook
+- Instagram
+- LinkedIn
+- [Microsoft identify platform](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow)
+
+# Custom Token
+
+In case the developer want to integrate with a provider who does not support OAuth 2.0, they can achieve single sign-on with Custom Token.
+
+### Client Side API
+
+```typescript
+function loginWithCustomToken(token: string, options?: LoginOptions): Promise<User>;
+```
+
+### Client Side Example
+
+```typescript
+loginMyAuthenticationServer().then(token => {
+  skygear.auth.loginWithCustomToken(token).then(user => {
+    // login suceeded
+  });
+});
+```
+
+### Server Side Integration
+
+The custom token **MUST** be a JWT token.
+
+It **MUST** have the following standard claims.
+
+- `iss`: The value must match the configured value in tenant configuration.
+- `sub`: The user ID of the foreign authentication system.
+
+It **SHOULD** have the following standard claims.
+
+- `aud`: If present and the configured value is not empty, they must match.
+- `exp`: If present, the Auth gear will validate it against the current time.
+- `nbf`: If present, the Auth gear will validate it against the current time.
+
+The signing algorithm is `HS256`. The signing key is shared between the Auth gear and the foreign authentication system.
+
+The token can optionally contain a claim named `skyprofile` which **MUST** be an JSON object.
+If `skyprofile` contains `email`, then merging applies.
+
+### Server Side Example (Node.js)
+
+```typescript
+// Install this package from npm
+const jwt = require("jsonwebtoken");
+
+const issuer = "MyAuthenticationSystem";
+const userID = "User ID of my foreign authentication system";
+const secret = "Shared secret with the Auth gear";
+const nowUnix = Math.floor(Date.now() / 1000);
+
+const token = jwt.sign({
+  sub: userID,
+  iss: issuer,
+  iat: nowUnix,
+  exp: nowUnix + 60 * 60,
+  skyprofile: {
+    email: "user@example.com",
+  },
+}, secret, { algorithm: "HS256" });
+```
+
+# Tenant Configuration
+
+```YAML
+url_prefix: ''
```

Comment:
Should this configurable by app developer? I think this URL would be set by the hosting admin.

Reply:
Same as `js_sdk_cdn_url`. Actually this setting should be put in `oauth`, let me correct this first.